import pdfplumber
import re
import os
import shutil
from template import render_filename_template, extract_placeholders, load_templates, save_templates, DEFAULT_TEMPLATES
from PySide6.QtWidgets import (
    QApplication, QMessageBox, QFileDialog, QMainWindow,
    QDialog, QListWidget, QPushButton, QComboBox, QLineEdit, QLabel,
    QVBoxLayout, QHBoxLayout, QGridLayout
)
from PySide6.QtUiTools import QUiLoader


def extract_info(text):
    name_match = re.search(r"æ—…å®¢å§“å.*?\n\s*([\S]+)", text)
    from_match = re.search(r"è‡ª[ï¼š:]\s*(\S+\s+\S+)", text)
    to_match = re.search(r"è‡³[ï¼š:]\s*(\S+\s+\S+)", text)
    amount_match = re.search(r"åˆè®¡.*?\n.*?(\d+\.\d{2})\s*$", text, re.MULTILINE)
    buyer_match = re.search(r"è´­ä¹°æ–¹åç§°ï¼š(\S+)", text)
    invoice_match = re.search(r"å‘ç¥¨å·ç [ï¼š:]\s*(\d+)", text)
    issue_date_match = re.search(r"å¡«å¼€æ—¥æœŸ[ï¼š:]\s*([0-9]{4}[å¹´/-][0-9]{1,2}[æœˆ/-][0-9]{1,2}[æ—¥]?)", text)

    name = name_match.group(1) if name_match else None
    origin = from_match.group(1) if from_match else None
    destination = to_match.group(1) if to_match else None
    amount = amount_match.group(1) if amount_match else None
    buyer = buyer_match.group(1) if buyer_match else None
    invoice_number = invoice_match.group(1) if invoice_match else None
    issue_date = issue_date_match.group(1) if issue_date_match else None

    return name, origin, destination, amount, buyer, invoice_number, issue_date


def safe_filename(s):
    return re.sub(r'[\\/:*?"<>|]', "_", s)


def get_unique_path(folder, base_name):
    base_name = safe_filename(base_name)
    full_path = os.path.join(folder, base_name)
    if not os.path.exists(full_path):
        return full_path

    name, ext = os.path.splitext(base_name)
    counter = 1
    while True:
        new_name = f"{name}_({counter}){ext}"
        new_path = os.path.join(folder, new_name)
        if not os.path.exists(new_path):
            return new_path
        counter += 1


class TemplateBuilderDialog(QDialog):
    def __init__(self, templates: dict, templates_path: str, parent=None):
        super().__init__(parent)
        self.setWindowTitle("æ¨¡æ¿åˆ¶ä½œ")
        self.templates_path = templates_path
        self.templates = dict(templates)
        
        self.setStyleSheet("""
            QDialog {
                background: #ffffff;
                font-family: "Segoe UI", "Microsoft YaHei", sans-serif;
                font-size: 15px;
            }
            
            QPushButton {
                background-color: #0ea5e9;
                color: #ffffff;
                border: none;
                border-radius: 8px;
                padding: 8px 20px;
                font-weight: 500;
                min-height: 16px;
            }
            QPushButton:hover {
                background-color: #0284c7;
            }
            QPushButton:pressed {
                background-color: #0369a1;
            }
            
            QLabel {
                color: #374151;
                font-weight: 500;
            }
            
            QComboBox {
                background: #ffffff;
                border: 1px solid #d1d5db;
                border-radius: 8px;
                padding: 8px 12px;
                min-height: 20px;
            }
            
            QLineEdit {
                background: #ffffff;
                border: 1px solid #d1d5db;
                border-radius: 8px;
                padding: 8px 12px;
                min-height: 20px;
            }
            
            QComboBox:hover, QLineEdit:hover {
                border: 1px solid #9ca3af;
            }
            
            QComboBox:focus, QLineEdit:focus {
                border: 1px solid #0ea5e9;
                outline: none;
            }
            
            QListWidget {
                background: #ffffff;
                border: 1px solid #e5e7eb;
                border-radius: 8px;
                padding: 4px;
            }
            
            QListWidget::item {
                padding: 6px;
                border-radius: 6px;
            }
            
            QListWidget::item:hover {
                background: #f3f4f6;
            }
            
            QListWidget::item:selected {
                background: #e5e7eb;
                color: #111827;
            }
        """)

        self.listTemplates = QListWidget()
        for name in self.templates.keys():
            self.listTemplates.addItem(name)
        self.listTemplates.itemClicked.connect(self._loadTemplate)

        # å³ä¾§ï¼šæ‹¼è£…å™¨
        self.fieldsCombo = QComboBox()
        # æ˜¾ç¤ºå‹å¥½ä¸­æ–‡ï¼Œæ˜ å°„ä¸ºå ä½ç¬¦key
        self.display_to_key = {
            "ä¹°å®¶": "buyer",
            "å§“å": "name",
            "å‡ºå‘åœ°": "origin",
            "ç›®çš„åœ°": "destination",
            "é‡‘é¢": "amount",
            "å¡«å¼€æ—¥æœŸ": "issue_date",
            "å‘ç¥¨å·": "invoice_number",
            "åŸæ–‡ä»¶å": "original_filename",
            "è‡ªå®šä¹‰æ–‡æœ¬": "__custom__",
        }
        for label in self.display_to_key.keys():
            self.fieldsCombo.addItem(label)

        self.customText = QLineEdit()
        self.customText.setPlaceholderText("è¾“å…¥è‡ªå®šä¹‰æ–‡æœ¬ï¼ˆé€‰æ‹©â€œè‡ªå®šä¹‰æ–‡æœ¬â€æ—¶å¯ç”¨ï¼‰")
        self.customText.setEnabled(False)
        self.fieldsCombo.currentTextChanged.connect(self._onFieldChange)

        self.btnAddSegment = QPushButton("+")
        self.btnAddSegment.clicked.connect(self._addSegment)
        self.btnRemoveLast = QPushButton("åˆ é™¤æœ€å")
        self.btnRemoveLast.clicked.connect(self._removeLast)
        self.btnClear = QPushButton("æ¸…ç©º")
        self.btnClear.clicked.connect(self._clearSegments)

        self.listSegments = QListWidget()
        self.lblPreview = QLabel("é¢„è§ˆï¼š")
        self.templateName = QLineEdit()
        self.templateName.setPlaceholderText("æ¨¡æ¿åç§°ï¼ˆå¦‚ï¼šå…¬å¸æ ¼å¼Aï¼‰")
        self.btnSaveTemplate = QPushButton("ä¿å­˜æ¨¡æ¿")
        self.btnSaveTemplate.clicked.connect(self._saveTemplate)
        self.btnClose = QPushButton("å…³é—­")
        self.btnClose.clicked.connect(self.close)

        self.btnNewTemplate = QPushButton("+")
        self.btnNewTemplate.clicked.connect(self._newTemplate)
        
        leftLayout = QVBoxLayout()
        leftTopLayout = QHBoxLayout()
        leftTopLayout.addWidget(QLabel("å½“å‰æ¨¡æ¿ä¸€è§ˆï¼š"))
        leftTopLayout.addStretch()
        leftTopLayout.addWidget(self.btnNewTemplate)
        leftLayout.addLayout(leftTopLayout)
        leftLayout.addWidget(self.listTemplates)

        assembleGrid = QGridLayout()
        assembleGrid.addWidget(QLabel("å…ƒç´ ï¼š"), 0, 0)
        assembleGrid.addWidget(self.fieldsCombo, 0, 1)
        assembleGrid.addWidget(QLabel("æ–‡æœ¬ï¼š"), 1, 0)
        assembleGrid.addWidget(self.customText, 1, 1)
        assembleGrid.addWidget(self.btnAddSegment, 0, 2)
        assembleGrid.addWidget(self.btnRemoveLast, 1, 2)
        assembleGrid.addWidget(self.btnClear, 2, 2)

        rightLayout = QVBoxLayout()
        rightLayout.addLayout(assembleGrid)
        rightLayout.addWidget(QLabel("ç‰‡æ®µåˆ—è¡¨ï¼š"))
        rightLayout.addWidget(self.listSegments)
        rightLayout.addWidget(self.lblPreview)
        rightLayout.addWidget(QLabel("æ¨¡æ¿åç§°ï¼š"))
        rightLayout.addWidget(self.templateName)
        rightLayout.addWidget(self.btnSaveTemplate)
        rightLayout.addWidget(self.btnClose)

        mainLayout = QHBoxLayout()
        mainLayout.addLayout(leftLayout, 1)
        mainLayout.addLayout(rightLayout, 2)
        self.setLayout(mainLayout)

        # å†…éƒ¨çŠ¶æ€ï¼šsegments ä¸ºå­—ç¬¦ä¸²ç‰‡æ®µï¼Œå¦‚ ["{buyer}", "--", "{amount}"]
        self.segments = []
        self._refreshPreview()

    def _newTemplate(self):
        self.listTemplates.clearSelection()
        self.segments = []
        self.listSegments.clear()
        self.templateName.clear()
        self._refreshPreview()
    
    def _onFieldChange(self, _text: str):
        is_custom = self.display_to_key.get(self.fieldsCombo.currentText()) == "__custom__"
        self.customText.setEnabled(is_custom)

    def _addSegment(self):
        display = self.fieldsCombo.currentText()
        key = self.display_to_key.get(display)
        if key == "__custom__":
            text = self.customText.text()
            if not text:
                QMessageBox.warning(self, "æç¤º", "è¯·è¾“å…¥è‡ªå®šä¹‰æ–‡æœ¬")
                return
            self.segments.append(text)
            self.listSegments.addItem(f"è‡ªå®šä¹‰æ–‡æœ¬ï¼š{text}")
        else:
            placeholder = "{" + key + "}"
            self.segments.append(placeholder)
            # åœ¨åˆ—è¡¨ä¸­æ˜¾ç¤ºä¸­æ–‡åç§°è€Œä¸æ˜¯å ä½ç¬¦
            self.listSegments.addItem(display)
        self._refreshPreview()

    def _removeLast(self):
        if self.segments:
            self.segments.pop()
            self.listSegments.takeItem(self.listSegments.count() - 1)
            self._refreshPreview()

    def _clearSegments(self):
        self.segments = []
        self.listSegments.clear()
        self._refreshPreview()

    def _refreshPreview(self):
        template_str = "".join(self.segments)
        self.lblPreview.setText(f"é¢„è§ˆï¼š{template_str}")

    def _loadTemplate(self, item):
        template_name = item.text()
        template_str = self.templates.get(template_name, "")
        if not template_str:
            return
        
        if template_str.lower().endswith('.pdf'):
            template_str = template_str[:-4]
        
        self.segments = []
        self.listSegments.clear()
        self.templateName.setText(template_name)
        
        i = 0
        while i < len(template_str):
            if template_str[i] == '{':
                end = template_str.find('}', i)
                if end != -1:
                    placeholder = template_str[i:end+1]
                    key = placeholder[1:-1]
                    for display, k in self.display_to_key.items():
                        if k == key:
                            self.segments.append(placeholder)
                            self.listSegments.addItem(display)
                            break
                    i = end + 1
                else:
                    self.segments.append(template_str[i])
                    self.listSegments.addItem(f"è‡ªå®šä¹‰æ–‡æœ¬ï¼š{template_str[i]}")
                    i += 1
            else:
                start = i
                while i < len(template_str) and template_str[i] != '{':
                    i += 1
                custom_text = template_str[start:i]
                if custom_text:
                    self.segments.append(custom_text)
                    self.listSegments.addItem(f"è‡ªå®šä¹‰æ–‡æœ¬ï¼š{custom_text}")
        
        self._refreshPreview()

    def _saveTemplate(self):
        name = self.templateName.text().strip()
        if not name:
            QMessageBox.warning(self, "æç¤º", "è¯·å¡«å†™æ¨¡æ¿åç§°")
            return
        template_str = "".join(self.segments)
        if not template_str:
            QMessageBox.warning(self, "æç¤º", "æ¨¡æ¿å†…å®¹ä¸ºç©º")
            return
        
        if not template_str.lower().endswith('.pdf'):
            template_str += '.pdf'
        
        self.templates[name] = template_str
        try:
            save_templates(self.templates_path, self.templates)
        except Exception as e:
            QMessageBox.critical(self, "é”™è¯¯", f"ä¿å­˜å¤±è´¥: {e}")
            return
        
        if not any(self.listTemplates.item(i).text() == name for i in range(self.listTemplates.count())):
            self.listTemplates.addItem(name)
        QMessageBox.information(self, "æˆåŠŸ", f"å·²ä¿å­˜æ¨¡æ¿ï¼š{name}")

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = uiLoader.load('main.ui')
        self.ui.setStyleSheet("""
            * {
                font-family: "Segoe UI", "Microsoft YaHei", sans-serif;
                font-size: 15px;
            }

            QMainWindow, QWidget {
                background: #ffffff;
            }

            QPushButton {
                background-color: #0ea5e9;
                color: #ffffff;
                border: none;
                border-radius: 8px;
                padding: 8px 20px;
                font-weight: 500;
                min-height: 16px;
            }
            QPushButton:hover {
                background-color: #0284c7;
            }
            QPushButton:pressed {
                background-color: #0369a1;
            }

            QGroupBox {
                border: 1px solid #e5e7eb;
                border-radius: 12px;
                margin-top: 12px;
                padding: 16px;
                background: #fafafa;
                font-weight: 600;
            }
            QGroupBox::title {
                subcontrol-origin: margin;
                left: 12px;
                padding: 0 8px;
                color: #111827;
                background: #fafafa;
                font-size: 16px;
            }

            QLabel {
                color: #374151;
                font-weight: 500;
            }

            QComboBox {
                background: #ffffff;
                border: 1px solid #d1d5db;
                border-radius: 8px;
                padding: 8px 12px;
                min-height: 20px;
            }
            
            QLineEdit {
                background: #ffffff;
                border: 1px solid #d1d5db;
                border-radius: 8px;
                padding: 8px 12px;
                min-height: 20px;
            }
            
            QComboBox:hover, QLineEdit:hover {
                border: 1px solid #9ca3af;
            }
            
            QComboBox:focus, QLineEdit:focus {
                border: 1px solid #0ea5e9;
                outline: none;
            }

            QTextBrowser {
                background: #f9fafb;
                color: #1f2937;
                border: 1px solid #e5e7eb;
                border-radius: 12px;
                padding: 12px;
                font-family: "Consolas", "Monaco", monospace;
                font-size: 14px;
                line-height: 1.6;
            }
            
            QListWidget {
                background: #ffffff;
                border: 1px solid #e5e7eb;
                border-radius: 8px;
                padding: 4px;
            }
            
            QListWidget::item {
                padding: 6px;
                border-radius: 6px;
            }
            
            QListWidget::item:hover {
                background: #f3f4f6;
            }
            
            QListWidget::item:selected {
                background: #e5e7eb;
                color: #111827;
            }
        """)
        self.setWindowOpacity(1)

        self.ui.btnGenerate.clicked.connect(self.generate)
        self.ui.btnSelectFolder.clicked.connect(self.selectFolder)
        if hasattr(self.ui, 'btnDesignTemplates'):
            self.ui.btnDesignTemplates.clicked.connect(self.openTemplateDesigner)
        self.import_folder = None
        self.templates_path = os.path.join(os.getcwd(), "templates.yaml")
        self.templates = {}
        # â€œè¯»å–æ¨¡æ¿â€æŒ‰é’®è§¦å‘è¯»å–/åˆ›å»º YAML å¹¶åˆ·æ–°ä¸‹æ‹‰
        if hasattr(self.ui, 'btnLoadTemplates'):
            self.ui.btnLoadTemplates.clicked.connect(self.loadTemplatesButton)
        # å¯åŠ¨æ—¶è¯»å–æ¨¡æ¿ä¸€æ¬¡
        self.loadTemplatesButton()
        

    def log(self, text):
        self.ui.textLog.append(text)
        print(text)

    def selectFolder(self):
        folder = QFileDialog.getExistingDirectory(None, "é€‰æ‹©PDFæ–‡ä»¶å¤¹")
        if folder:
            self.import_folder = folder
            self.log(f"ğŸ“ å·²é€‰æ‹©æ–‡ä»¶å¤¹: {folder}")
        else:
            self.log("âš ï¸ æœªé€‰æ‹©æ–‡ä»¶å¤¹")

    # Template management helpers
    def refreshTemplateCombo(self):
        try:
            self.ui.cmbTemplate.clear()
            for name in self.templates.keys():
                self.ui.cmbTemplate.addItem(name)
            # å¦‚æœæ²¡æœ‰ä»»ä½•æ¨¡æ¿ï¼Œæ·»åŠ é»˜è®¤
            if self.ui.cmbTemplate.count() == 0:
                for name in DEFAULT_TEMPLATES.keys():
                    self.ui.cmbTemplate.addItem(name)
                self.templates = DEFAULT_TEMPLATES.copy()
        except Exception as e:
            self.log(f"âš ï¸ åˆ·æ–°æ¨¡æ¿ä¸‹æ‹‰å¤±è´¥: {e}")

    def loadTemplatesButton(self):
        # ç‚¹å‡»æŒ‰é’®æ—¶è¯»å– YAMLï¼›è‹¥ä¸å­˜åœ¨åˆ™åˆ›å»ºé»˜è®¤ YAML åå†è¯»å–
        if not os.path.exists(self.templates_path):
            try:
                save_templates(self.templates_path, DEFAULT_TEMPLATES)
                self.log(f"âœ… å·²åˆ›å»ºé»˜è®¤æ¨¡æ¿æ–‡ä»¶: {self.templates_path}")
            except Exception as e:
                self.log(f"âš ï¸ åˆ›å»ºé»˜è®¤æ¨¡æ¿æ–‡ä»¶å¤±è´¥: {e}")
        self.templates = load_templates(self.templates_path)
        self.refreshTemplateCombo()
        self.log("âœ… æ¨¡æ¿å·²è¯»å–å¹¶æ›´æ–°ä¸‹æ‹‰é€‰é¡¹")

    def openTemplateDesigner(self):
        # ç¡®ä¿æœ‰ YAML æ–‡ä»¶
        if not os.path.exists(self.templates_path):
            try:
                save_templates(self.templates_path, DEFAULT_TEMPLATES)
                self.log(f"âœ… å·²åˆ›å»ºé»˜è®¤æ¨¡æ¿æ–‡ä»¶: {self.templates_path}")
            except Exception as e:
                self.log(f"âš ï¸ åˆ›å»ºé»˜è®¤æ¨¡æ¿æ–‡ä»¶å¤±è´¥: {e}")
        # è¯»å–ç°æœ‰æ¨¡æ¿ï¼Œæ‰“å¼€å¯¹è¯æ¡†
        self.templates = load_templates(self.templates_path)
        dlg = TemplateBuilderDialog(self.templates, self.templates_path, self)
        dlg.exec()
        # é‡æ–°åŠ è½½ï¼ˆè‹¥å¯¹è¯æ¡†æœŸé—´æœ‰å˜åŒ–ï¼‰å¹¶åˆ·æ–°ä¸‹æ‹‰
        self.templates = load_templates(self.templates_path)
        self.refreshTemplateCombo()

    def generate(self):
        if not self.import_folder:
            QMessageBox.warning(None, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©ä¸€ä¸ªå¯¼å…¥æ–‡ä»¶å¤¹")
            return

        import_folder = self.import_folder
        export_folder = os.path.join(import_folder, "exports")
        unsure_folder = os.path.join(export_folder, "unsure_folder")

        os.makedirs(export_folder, exist_ok=True)
        os.makedirs(unsure_folder, exist_ok=True)

        for file_name in os.listdir(import_folder):
            if file_name.lower().endswith(".pdf"):
                file_path = os.path.join(import_folder, file_name)
                try:
                    with pdfplumber.open(file_path) as pdf:
                        full_text = ""
                        for page in pdf.pages:
                            page_text = page.extract_text()
                            if page_text:
                                full_text += page_text + "\n"

                    name, origin, destination, amount, buyer, invoice_number, issue_date = extract_info(full_text)
                    self.log(f"ğŸ” æå–ä¿¡æ¯: {file_name} -> {name}, {origin}, {destination}, {amount}, {buyer}, å‘ç¥¨å·ç : {invoice_number}, å¡«å¼€æ—¥æœŸ: {issue_date}")

                    try:
                        selected_template_name = self.ui.cmbTemplate.currentText()
                    except Exception:
                        selected_template_name = "è¡Œç¨‹ä¿¡æ¯"

                    # åŠ¨æ€å­—æ®µæ ¡éªŒï¼šæ ¹æ®æ¨¡æ¿å†³å®šå¿…éœ€å­—æ®µ
                    values = {
                        "name": name,
                        "origin": origin.replace(' ','') if origin else None,
                        "destination": destination.replace(' ','') if destination else None,
                        "amount": amount,
                        "buyer": buyer,
                        "invoice_number": invoice_number,
                        "issue_date": issue_date,
                        "original_filename": os.path.splitext(file_name)[0],
                    }

                    template_str = self.templates.get(selected_template_name, DEFAULT_TEMPLATES["è¡Œç¨‹ä¿¡æ¯"])
                    required_fields = list(extract_placeholders(template_str))

                    if any(values.get(field) in (None, "") for field in required_fields):
                        dest_path = get_unique_path(unsure_folder, file_name)
                        shutil.copy(file_path, dest_path)
                        self.log(f"âš ï¸ ä¿¡æ¯ä¸å…¨ï¼ˆç¼ºå°‘: {', '.join([f for f in required_fields if not values.get(f)])}ï¼‰ï¼Œå·²ç§»åŠ¨è‡³ unsure_folder: {os.path.basename(dest_path)}")
                        continue

                    new_name = render_filename_template(template_str, values)
                    if not new_name.lower().endswith('.pdf'):
                        new_name += '.pdf'
                    dest_path = get_unique_path(export_folder, new_name)
                    shutil.copy(file_path, dest_path)
                    self.log(f"âœ… å·²ä¿å­˜: {os.path.basename(dest_path)}")

                except Exception as e:
                    self.log(f"âŒ å¤„ç†å‡ºé”™: {file_name}, é”™è¯¯: {type(e).__name__}: {e}")


if __name__ == "__main__":
    uiLoader = QUiLoader()
    app = QApplication([])

    stats = MainWindow()
    stats.ui.show()

    app.exec()

# pyinstaller --onefile --windowed --icon=main.ico main.py